test/Microsoft.AspNetCore.Razor.LanguageServer.Test/DefaultTagHelperDescriptionFactoryTest.cs

```diff
@@ -0,0 +1,381 @@
+        {
+            // Arrange
+            var content = "Microsoft.AspNetCore.SometTagHelpers.SomeType<Foo.Bar<Baz.Phi>>";
+
+            // Act
+            var reduced = DefaultTagHelperDescriptionFactory.ReduceTypeName(content);
+
+            // Assert
+            Assert.Equal("SomeType<Foo.Bar<Baz.Phi>>", reduced);
+        }
+
+        [Theory]
+        [InlineData("Microsoft.AspNetCore.SometTagHelpers.SomeType.Foo.Bar<Baz.Phi>>")]
+        [InlineData("Microsoft.AspNetCore.SometTagHelpers.SomeType.Foo.Bar{Baz.Phi}}")]
+        public void ReduceTypeName_UnbalancedDocs_NotRecoverable_ReturnsOriginalContent(string content)
+        {
+            // Arrange
+
+            // Act
+            var reduced = DefaultTagHelperDescriptionFactory.ReduceTypeName(content);
+
+            // Assert
+            Assert.Equal(content, reduced);
+        }
+
+        [Fact]
+        public void ReduceMemberName_Plain()
+        {
+            // Arrange
+            var content = "Microsoft.AspNetCore.SometTagHelpers.SomeType.SomeProperty";
+
+            // Act
+            var reduced = DefaultTagHelperDescriptionFactory.ReduceMemberName(content);
+
+            // Assert
+            Assert.Equal("SomeType.SomeProperty", reduced);
+        }
+
+        [Fact]
+        public void ReduceMemberName_Generics()
+        {
+            // Arrange
+            var content = "Microsoft.AspNetCore.SometTagHelpers.SomeType<Foo.Bar>.SomeProperty<Foo.Bar>";
+
+            // Act
+            var reduced = DefaultTagHelperDescriptionFactory.ReduceMemberName(content);
+
+            // Assert
+            Assert.Equal("SomeType<Foo.Bar>.SomeProperty<Foo.Bar>", reduced);
+        }
+
+        [Fact]
+        public void ReduceMemberName_CrefGenerics()
+        {
+            // Arrange
+            var content = "Microsoft.AspNetCore.SometTagHelpers.SomeType{Foo.Bar}.SomeProperty{Foo.Bar}";
+
+            // Act
+            var reduced = DefaultTagHelperDescriptionFactory.ReduceMemberName(content);
+
+            // Assert
+            Assert.Equal("SomeType{Foo.Bar}.SomeProperty{Foo.Bar}", reduced);
+        }
+
+        [Fact]
+        public void ReduceMemberName_NestedGenericsMethodsTypes()
+        {
+            // Arrange
+            var content = "Microsoft.AspNetCore.SometTagHelpers.SomeType<Foo.Bar<Baz,Fi>>.SomeMethod(Foo.Bar<System.String>,Baz<Something>.Fi)";
+
+            // Act
+            var reduced = DefaultTagHelperDescriptionFactory.ReduceMemberName(content);
+
+            // Assert
+            Assert.Equal("SomeType<Foo.Bar<Baz,Fi>>.SomeMethod(Foo.Bar<System.String>,Baz<Something>.Fi)", reduced);
+        }
+
+        [Theory]
+        [InlineData("Microsoft.AspNetCore.SometTagHelpers.SomeType.Foo.Bar<Baz.Phi>>")]
+        [InlineData("Microsoft.AspNetCore.SometTagHelpers.SomeType.Foo.Bar{Baz.Phi}}")]
+        [InlineData("Microsoft.AspNetCore.SometTagHelpers.SomeType.Foo.Bar(Baz.Phi))")]
```

Comment:
Add more malformed cases here. Like `Foo..Bar`, `Foo{.>` etc. There are probably not realistic scenarios but doesn't hurt to make sure we don't explode.

Reply:
Can add your second scenario. The first one we'd actually return a reduced member because we don't care if the type name is string.Empty. (multiple dots are recoverable)